### ISSUE TITLE
Add support for "here documents"

### ISSUE THREAD

**Issue by gustafj** [fish-shell/fish-shell#540]
This is a feature request for "here documents" (using ex bash's syntax, or improving on it).

Example:
```
<< EOF
This is a here document,
that spans multiple lines
EOF
```

Or:
```
<<
Here document # 2
ctrl+d
```

I use this quite frequently when copy-pasting to a terminal input which i want to modify somewhat with standard tools instead of saving to a temp file first.

See: http://www.gnu.org/software/bash/manual/bashref.html#Here-Documents
Or: http://en.wikipedia.org/wiki/Here_document

**Comment by metal3d** [comment]
:+1: - I just wonder how to user pseudo file too... At this time I really need to find a way to make

```
# bash equivalent:
# triple sign to simulate file
command <<< "abc"
# or
command <(echo "abc")

# equivalent to 
echo "abc" > /tmp/test
command /tmp/test
```

At this time I really don't find any solution. A lot of command doesn't read STDIN and wait a real file as argument. Triple sign or "command as file with <(...)" is, IMHO, a necessity

**Comment by scooter-dangle** [comment]
Not positive it would be everything you're looking for, but have you tried
```fish
command (echo "abc" | psub)
```
?

**Comment by metal3d** [comment]
@scooter-dangle god, I didn't know this command. Very nice ! thanks !

To be precised, when I say that I need herestring or heredoc, I mean that I need solutions as the one you give. I just need solution, syntax is not the problem. And "psub" is a very good solution for my problem.

I guess that I will create a last ticket for my last need that is the >(cmd) syntax. An example:

```
exec 3> >(zenity --progress )

for i in $(seq 100); do
    sleep 0.5
    echo $i
    echo $i >&3
done
```
That is another ticket, I know :)

**Comment by zanchey** [comment]
Indeed, see #1786.

**Comment by metal3d** [comment]
@zanchey tried, didn't work... I will comment the issue you pointed, thanks a lot

**Comment by howdoicomputer** [comment]
I just make calls to bash when I want to use POSIX 'features.' I do wish there was a better solution, however.

```bash
➜  workspace bash -c "cat << 'EOF' | drafter --format yaml"
_version: "3.0"
metadata: []
name: ""
description: ""
element: ""
resourceGroups: []
content: []

OK.
```

**Comment by zanchey** [comment]
That's a good example of something that doesn't need bash (or cat): you can just run drafter directly and paste your text straight in, then hit Ctrl-D to close the stream.

**Comment by faho** [comment]
So far, nobody has given any example that can't be solved by eliminating `cat` and pasting or using `echo` with multi-line literals, so I'm closing this.

If anybody has an example where that isn't a solution, please comment!

**Comment by dxlr8r** [comment]
I think fish needs this. I've read all the comments and while echo/printf | read etc. all work to some degree it doesn't handle quotations very good. 

If you want to store some commands in a string, say some commands you send through ssh to execute on a remote computer you would have to escape the outer quotation in the inner commands you want to run. Like in this case where the remote host uses bash/sh:

`echo "a=b; echo \"Var a is $a\"" > read foo`

Ofcourse you could wrap it in ', but you would still run into troubles where you need ' on the remote (for not to parse $ in grep which is used to mark the end of a line): 

`echo 'echo "abc" | grep \'^abc$\'' > read foo`

It gets very complicated in the long run. Especially if you nest commands inside commands `ssh foo "sh -c 'sh -c \"\"'"` and such.

**Comment by dxlr8r** [comment]
Here is my attempt at some kind of here doc in Fish. Works in script files.

```
function heredoc
    set label $argv[1]
    set script (status -f)
    set print -1

    for line in (cat $script)
       	if test (echo $line | grep '^#'$label)  
       	    # print from, not including, label to label
            set print (math "$print*-1")
       	else if test $print -eq 1
            echo $line | sed 's/^.//' # remove first character (#)
       	end
    end
end
```

Then you could do inside a script file:

```
#foo
#ls -la ""
#pwd
#foo

echo (heredoc foo) | read foobar
```

Now $foobar would contain everything between #foo and #foo with the first character (#) removed.

> ls -la ""
> pwd

**Comment by krader1961** [comment]
@dxlr8r, That is a creative solution but is not generally applicable. The only way for a "here doc" to work is via special syntax recognized by the shell language parser. Your attempt at a solution via a shell function will not work except in an extremely narrow set of circumstances. Which means your solution should not be part of the core fish code.

**Comment by ridiculousfish** [comment]
That's like LISP-level clever. Well done.

**Comment by dxlr8r** [comment]
I agree @krader1961. I just wanted to share it, maybe it's useful for someone until real support is implemented. 

@ridiculesfish thanks, Clojure is my #1 language. Might be the reason :)

**Comment by burke** [comment]
This is something I'd really love to see a fix for.

Most of the bash/zsh features absent in fish are easy to find an alternate pattern for, but lack of heredocs makes it difficult to embed an arbitrary file in a fish script. 

---

Here are some overly specific details on my actual use-case:

<details>

I have some software that installs a few functions into the user's shell. I want to serve up the installer as a single script.

* Because I want to re-exec the user's shell to load the functions at the end, I need to run the script with `eval`.
* Because people have crazy shell configuration, and I support bash, zsh, and fish, I want to run the meat of the script in a non-interactive context without as much shell configuration, and written once instead of 2-3 times.

To accommodate this, I have a `stage1.sh` and `stage2.sh`. `stage1.sh` writes `stage2.sh` out to a separate file and executes it with bash, then re-execs the user shell. `stage2.sh` does lots of stuff.

```bash
#!/bin/bash
# ^ this is for shellcheck, but in reality this is just executed by whatever
# the user's shell is.

# We're running in a highly-unpredictable environment here:
#
# For ZSH users with the CLOBBER option unset, > will fail if the file exists,
# and mktemp creates the file. By removing it first, we play it safe.
#
# By fully specifying paths to binaries and builtin for builtins, we can
# hopefully work around any particularly weird user configuration.

tempfile="$(/usr/bin/mktemp /tmp/dev-bootstrap-stage2-XXXXXX)"
/bin/rm -f "${tempfile}"

# The Makefile templates stage2.sh into the heredoc here.
/bin/cat << 'EOF' > "${tempfile}"
##STAGE2##
EOF

if builtin test $? -ne 0; then
  builtin echo "couldn't write bootstrap stage 2!" >&2
  /bin/rm -f "${tempfile}"
  builtin false
else
  /bin/bash "${tempfile}"
  if builtin test $? -eq 0; then
    /bin/rm -f "${tempfile}"
    # re-exec the user's shell to pick up the new `dev` function.
    builtin exec "$(/usr/bin/dscl . -read "/Users/$(/usr/bin/whoami)" UserShell | /usr/bin/awk '{print $NF}')" --login
  else
    /bin/rm -f "${tempfile}"
    builtin false
  fi
fi
```

...and the `Makefile`, fwiw:

```make
site/up: bootstrap/stage1.sh bootstrap/stage2.sh
        @mkdir -p $(@D)
        /usr/bin/sed -e '/##STAGE2##/ {' -e 'r $(word 2,$^)' -e d -e '}' < $< > $@
        chmod 755 $@
```

The general problem is that I can't take that `stage2.sh` file, embed it into a fish script without additional translation, and be able to write it out to a file.
</details>

**Comment by faho** [comment]
>Because I want to re-exec the user's shell to load the functions at the end, I need to run the script with eval

Why not just copy the functions into $fish_function_path? That way all currently running shells will pick them up.

**Comment by faho** [comment]
># For ZSH users with the CLOBBER option unset, > will fail if the file exists,
# and mktemp creates the file. By removing it first, we play it safe.

I'm quite certain zsh also knows the `>|` syntax to always clobber.

**Comment by burke** [comment]
> I'm quite certain zsh also knows the >| syntax to always clobber.

Oh, TIL. Thanks!

> Why not just copy the functions into $fish_function_path? That way all currently running shells will pick them up.

There's a little more ceremony to it than that. There's a bit of code I want to run on shell init too, and to keep the divergence low between the bash/zsh and fish paths, I'm just going to change my makefile to embed the other file in a string. 

Anyway, I can solve this instance of the problem easily enough -- it's just a bit awkward, and I figured I'd drop it here as an example of something that's awkward without heredocs -- but that embedded file could be anything within the general theme of "I'd like to serve up a single-file fish script that runs some additional file with some other interpreter". That use-case is currently much more awkward in fish than in other shells. 

(I don't need a resolution; just providing a data point)

**Comment by lloeki** [comment]
One typical requirement of herestrings and one way to achieve this in bash is [presented here](https://github.com/lloeki/digitalocean-debian-to-arch/blob/debian8/install.sh). Look for ` `extract_embedded_file()` and stuff after `cat <<<'EMBEDDED'`. I believe @dxlr8r gets close to this, but it may or may not always work on corner cases as it relies on `status -f` (does it work when sourced? symlinks?). And it requires a prefix before copy/pasting the body.

Could it be possible to have something akin to Perl's `__END__` and/or `__DATA__` [special tokens](http://perldoc.perl.org/perldata.html#Special-Literals). Even better if we can have the equivalent of `PACKNAME::DATA`. Or should we hack something with `^D` or `^Z`?

Unfortunately, this results in the following (that's a raw, literal `^D` char in the middle):

```
$ cat foo.fish
#!/usr/bin/env fish

echo 'noop'

^D
lorem ipsum dolor sit amet
$ fish foo.fish 
noop
fish: Unknown command ''
./foo.fish (line 5): 
                     ^
fish: Unknown command 'lorem ipsum dolor sit amet'
./foo.fish (line 6): lorem ipsum dolor sit amet
```

Here's an alternative, but it requires an `exit` before `__DATA__`:

```
#!/usr/bin/env fish

function read_data
    perl -ne 'print if $p==1; /^__DATA__$/ and $p=1' (status -f)
end

echo '==== BEGIN DATA ===='
read_data
echo '==== END DATA ===='

exit
__END__
some data
```

One could get creative and match only on `/__DATA__$/`, so that the final marker would be:

```
exit # __DATA__
```

But the whole point of `__DATA__` in Perl and the use of `cat << 'EMBEDDED'` in bash is to *not* have a call to `exit` because sometimes you want to source the file and not have the sourcer exit.

**Comment by faho** [comment]
>One typical requirement of herestrings and one way to achieve this in bash is presented here. Look for extract_embedded_file() and stuff after cat <<<'EMBEDDED'.

Okay, what this does is embed a massive string that contains multiple different entries that go to different files. Which part goes where is specified via '!!!file' beginning and '!!!' end markers.

So, my question is still:

How is this different from just embedding the string?

E.g. instead of

```fish
cat <<EMBEDDED
!!!!digitalocean-synchronize.service
[Unit]
Description=DigitalOcean Synchronization
DefaultDependencies=no
Before=systemd-networkd.service
After=systemd-udevd.service
[Service]
Type=oneshot
ExecStart=/usr/sbin/digitalocean-synchronize
!!!!
EMBEDDED
```

and then running an awk script on this, why not

```fish
set -l service '[Unit]
Description=DigitalOcean Synchronization
DefaultDependencies=no
Before=systemd-networkd.service
After=systemd-udevd.service
[Service]
Type=oneshot
ExecStart=/usr/sbin/digitalocean-synchronize'
echo $service > digitalocean-synchronize.service
```

Sure, you need to escape single quotes here, but that's kinda easy - `s/'/\\'/g`.

This has the advantage that it just reuses concepts that we have anyway.

Introducing new concepts should be done with care. It should be appreciably different from those we already have, and should enable something interesting. That's why we don't have `echo ${var:-something}` and such, because it's easy enough to use `set -q var; and echo $var; or echo something`.

The aim is not to allow code that uses the fewest possible characters, it's to allow code that is easy to read, and one way of achieving that is to cut down on the number of concepts.

**Comment by lsvermeer** [comment]
Correct me if I'm wrong, but I think here docs / pseudofiles are necessary to redirect from stdin to a program inside gdb (when using fish shell).

in bash:
    gdb ./someprogram
    (gdb) run <<< $(python -c 'print("\x00"*15 + "\n" + "...")')

in fish:
    gdb ./someprogram
    (gdb) run ???

**Comment by zanchey** [comment]
That doesn't require here documents (but does assume that your SHELL is set to fish). Instead, it requires process substitution to be translated from sh syntax to fish syntax:

```
gdb ./someprogram
(gdb) run < (python -c 'print("\x00"*15 + "\n" + "...")' | psub)
```

(Caveats on psub apply ie #1040.)

**Comment by eigenlambda** [comment]
What people here asking for here documents probably want is
set -h EOF -i heredoc
  "complicated language here uses all the quote's"
  with optional stripped `indentation
EOF
echo $heredoc

**Comment by dscottboggs** [comment]
@eigenlambda your solution looks like it would achieve the expected goal, but it doesn't seem to be working. The `-h` option displays the help message.

**Comment by nertpinx** [comment]
What @eigenlambda mentioned could help.  What I am using in scripts and missing in fish (or do not know how to do) is something along the lines of:

``` bash
gdb -x /proc/self/fd/5 -p $some_pid 5<<EOF
break func if param == 0
commands
break subfunc
continue
end
EOF
```

Of course it also uses an extra feature of redirecting the input to yet another FD, which is another thing I'm trying to figure out how to do with fish, but that is orthogonal to this issue.

For more convoluted example, this is from a script that sets up my whole router without transferring files or anything like that (effectively layman's ansible replacement):

``` bash
add_firewall() {
    NAME="$1"
    PORT="$2"
    PROTO="$3"
    SRC="$4"
    TARGET="$5"

    if [[ -z "$NAME" || -z "$PORT" || -z "$PROTO" || -z "$SRC" || -z "$TARGET" ]]; then
        echo "Missing argument" >&2
        exit 1
    fi

    ${SSH} "${ROUTER_CONN}" <<EOF
    set -e
    ${DEBUG}

    uci batch <<UCI_EOF
set firewall.${NAME}=rule
set firewall.${NAME}.dest_port='${PORT}'
set firewall.${NAME}.enabled=1
set firewall.${NAME}.name='${NAME}'
set firewall.${NAME}.proto='${PROTO}'
set firewall.${NAME}.src='${SRC}'
set firewall.${NAME}.target='${TARGET}'
UCI_EOF
EOF
}
```

It would be really nice to use some advantages of fish here.

**Comment by faho** [comment]
@nertpinx The gdb thing does not actually need an fd at all. Just put the commands into a file (or `echo` them into a tempfile) and pass the path to that instead of the path to the fd. With the firewall thing just use a pipe with e.g printf.

**Comment by lilyball** [comment]
For the gdb thing you can even `echo` them into `psub` in a command substitution, and it will synthesize the temporarily file for you and clean it up when the job completes.

**Comment by lloeki** [comment]
> Just put the commands into a file

That sort of defeats the purpose of heredocs, which in some use cases is to not have content end up on the filesystem.

**Comment by nertpinx** [comment]
@faho Creating a file does not work when you are e.g. suggesting a command for someone to run or supplying a script that could just work, but now would have to deal with files. I'm not saying it is not possible, it's just very unfriendly when you need to deal with way more than you are trying to solve.
@lilyball Yes, but again compare this:
```
<<EOF
this
is
a
$file
EOF
```
to this:
```
<(begin
echo this
echo is
echo a
echo "$file"
end | psub)
```
for more complicated examples it becomes a PITA to write (and escape properly sometimes).

**Comment by chengxuncc** [comment]
I must to say that it is a necessary feature for scripts, which can be simple and useful.
I place my here document into function, and I really want to implement it without beginning of `# `, but I can't, because it's impossible under fish rules.
```
#!/usr/bin/fish

function mydoc
revesed or something else
# every line of document
# should
# start at index 3 and begin with #
# ~!@#$%^&*()_+`1234567890-={}|[]\:";',./<>?
end

function heredoc -d "keep my hand from dirty work"
  set codefun (functions $argv[1])
  switch $codefun[1]
  case '#*'
    set codefun $codefun[4..-2]
  case '*'
    set codefun $codefun[3..-2]
  end
  string sub -s 3 $codefun
end

heredoc mydoc
``` 
[similar implement](https://github.com/fish-shell/fish-shell/issues/540#issuecomment-239155879)

**Comment by filiden** [comment]
I believe we can implement "here-document" transparently for the user.

> The redirection operators "<<" and "<<-" both allow redirection of subsequent lines read by the shell to the input of a command. The redirected lines are known as a "here-document".

Maybe you could *emulate* "here-document". Below, a [concrete exemple](https://stackoverflow.com/questions/35365769/how-do-i-use-jq-to-convert-number-to-string/35365770#35365770).

POSIX compatible way:

    jq -r '.[] | [ .string, .number|tostring ] | join(": ")' <<EOF
    [{ "number": 9, "string": "nine"},
     { "number": 4, "string": "four"}]
    EOF

Using Fish it would be:

    jq -r '.[] | [ .string, .number|tostring ] | join(": ")' -
    [{ "number": 9, "string": "nine"},
     { "number": 4, "string": "four"}]

It works but it seems that *we cannot edit the input text easily* probably because the input stream is managed by the application. With Bash using a "here-document", you can move the cursor using several keys.

Therefore, the idea would be to manage the input text on the command line "like intrinsically" a "here-document" rather than by the application. Nonetheless, I have no idea if it may break some features or introduce weird bugs.

In conclusion, Fish could use the symbol `-` as a here-document. The user enters the input text on a newline and validates its input using e.g. <kbd>Control+D</kbd>. We bypass the standard operation (or in other words, we adapt it if lucky).

By the way, I disagree with the following statement "Here documents are too similar to using echo inside of a pipeline."

**Comment by lilyball** [comment]
Fish cannot use `-` to start a here document, that would break an extremely large number of programs that use `-` to signal that they should read from standard input instead of from a file.

**Comment by keyone** [comment]
Try
```
printf "Hello World" >> hello.txt
```
Then
```
cat hello.txt
```

**Comment by luxlogica** [comment]
Unfortunately, the link to the original document posted by @xfix is now dead, so there is no way for us to understand why heredoc is not present in fish, or what is the 'standard' way to go about replacing that functionality. I've found that in the bash world, heredoc is often used when writing scripts where the user wants to connect via ssh to a remote machine, and run several commands on the remote - something like:

```
ssh user@remote.com 'bash -s' <<< EOF
sudo apt update
sudo apt upgrade -y
sudo apt auto remove
EOF
```

This example is trivial, and could be replaced by a long single-line command, but you get the point: image a more complex script, with dozens and dozens of lines, and you can see how the syntax does seem useful - both for maintainability and code clarity. Even in other shells, the workarounds to using heredoc usually involve splitting the code into multiple files (e.g., cat myscript.sh | ssh), struggling with quotes (echo "'$var'") which don't always work, or writing code that has to pipe itself through several tools (and becomes harder to maintain and understand). 

If the heredoc syntax is indeed missing _by design_, I'm guessing that there must be an 'official' workaround, or replacement - fish always seems to do things the 'smarter' way. What is it, then? It might be useful to have it spelled out here, or even better, **added to the official online guide**, so that others might find it more easily.

**Comment by lilyball** [comment]
The original link can be read at https://web.archive.org/web/20140316225710/http://ridiculousfish.com/shell/user_doc/html/design.html. It says

> Here documents are too similar to using echo inside of a pipeline.

This does not seem like a compelling argument to me.
